go/oasis-node/cmd/config/migrate/migrate.go

```diff
@@ -0,0 +1,617 @@
+							delete(m(m(tendermint)["core"]), "listen_address")
+						}
+						if ea, ok := m(v)["external_address"]; ok {
+							logger.Info("consensus.tendermint.core.external_address is now consensus.external_address")
+							m(newCfg["consensus"])["external_address"] = ea
+							delete(m(m(tendermint)["core"]), "external_address")
+						}
+					} else if k == "log" {
+						if dbg, ok := m(v)["debug"]; ok {
+							logger.Info("consensus.tendermint.log.debug is now consensus.log_debug")
+							m(newCfg["consensus"])["log_debug"] = dbg
+							delete(m(m(tendermint)["log"]), "debug")
+						}
+					} else if k == "light_client" {
+						if tp, ok := m(v)["trust_period"]; ok {
+							logger.Info("consensus.tendermint.light_client.trust_period is now consensus.state_sync.trust_period")
+							mkSubMap(m(newCfg["consensus"]), "state_sync")
+							m(m(newCfg["consensus"])["state_sync"])["trust_period"] = tp
+							delete(m(m(tendermint)["light_client"]), "trust_period")
+						}
+					} else if k == "seed" {
+						if dbg, ok := m(v)["debug"]; ok {
+							if dabfg, ok := m(dbg)["disable_addr_book_from_genesis"]; ok {
+								logger.Info("consensus.tendermint.seed.debug.disable_addr_book_from_genesis is now consensus.debug.disable_addr_book_from_genesis")
+								mkSubMap(m(newCfg["consensus"]), "debug")
+								m(m(newCfg["consensus"])["debug"])["disable_addr_book_from_genesis"] = dabfg
+								delete(m(m(m(tendermint)["seed"])["debug"]), "disable_addr_book_from_genesis")
+							}
+						}
+					} else if k == "sentry" {
+						if ua, ok := m(v)["upstream_address"]; ok {
+							logger.Info("consensus.tendermint.sentry.upstream_address is now consensus.sentry_upstream_addresses")
+							m(newCfg["consensus"])["sentry_upstream_addresses"] = ua
+							delete(m(m(tendermint)["sentry"]), "upstream_address")
+						}
+					} else if k == "upgrade" {
+						if sd, ok := m(v)["stop_delay"]; ok {
+							logger.Info("consensus.tendermint.upgrade.stop_delay is now consensus.upgrade_stop_delay")
+							m(newCfg["consensus"])["upgrade_stop_delay"] = sd
+							delete(m(m(tendermint)["upgrade"]), "stop_delay")
+						}
+					} else if k == "supplementarysanity" {
+						logger.Info("consensus.tendermint.supplementarysanity.* is now consensus.supplementary_sanity.*")
+						m(newCfg["consensus"])["supplementary_sanity"] = v
+						delete(m(tendermint), k)
+					} else if k == "p2p" {
+						mkSubMap(m(newCfg["consensus"]), "p2p")
+						for pk, pv := range m(v) {
+							nk := pk
+							if pk == "persistent_peer" {
+								logger.Info("consensus.tendermint.p2p.persistent_peer is now consensus.p2p.persistent_peers")
+								nk = "persistent_peers"
+							} else if pk == "unconditional_peer" {
+								logger.Info("consensus.tendermint.p2p.unconditional_peer is now consensus.p2p.unconditional_peers")
+								nk = "unconditional_peers"
+							} else if pk == "seed" {
+								logger.Info("consensus.tendermint.p2p.seed is now p2p.seeds")
```

Comment:
Note that the previous version was a bit different -- it took a Tendermint P2P address where it now takes a P2P public key before the `@` part. Unfortunately, these cannot be migrated that easily as the hash function is one-way.

Either we need to support both (in `ConsensusAddressesToTendermint`) or we need to throw an error/warning here (something that will not get lost with all of the other info statements) that this will not work directly.

Reply:
Agreed.

Reply:
Hmm, good catch, I will add an error here.

Reply:
I've made it emit an error here, as well as leave the `tendermint.p2p.seed` section unmigrated, so it gets printed at the end of the process, which is more noticeable to the user.